util/expensivequery/expensivequery.go

```diff
@@ -73,12 +85,214 @@ func (eqh *Handle) Run() {
+			zap.Any("serverMemoryQuota", eqh.record.serverMemoryQuota),
+		)
+	} else {
+		logutil.BgLogger().Warn("The os system now takes a lot of memory, has the risk of OOM",
+			zap.Any("osMemUsage", memUsage),
+			zap.Any("osMemoryTotal", eqh.record.serverMemoryQuota),
+		)
+	}
+
+	if eqh.record.err = disk.CheckAndInitTempDir(); eqh.record.err != nil {
+		return
+	}
+	eqh.oomRecordSQL()
+	eqh.oomRecordProfile()
+
+	tryRemove := func(filename []string) {
+		// Keep the last 5 files
+		if len(filename) < 6 {
+			return
+		}
+		err := os.Remove(filename[0])
+		if err != nil {
+			logutil.BgLogger().Error("Remove temp files failed.", zap.Error(err))
+		}
+		filename = filename[1:]
+	}
+	tryRemove(eqh.record.lastLogFileName)
+	for i := range eqh.record.lastProfileFileName {
+		tryRemove(eqh.record.lastProfileFileName[i])
+	}
+}
+
+func (eqh *Handle) oomRecordSQL() {
+	sm := eqh.sm.Load().(util.SessionManager)
+	processInfo := sm.ShowProcessList()
+	pinfo := make([]*util.ProcessInfo, 0, len(processInfo))
+	for _, info := range processInfo {
+		if len(info.Info) != 0 {
+			pinfo = append(pinfo, info)
+		}
+	}
+	now := time.Now()
+
+	fileName := filepath.Join(config.GetGlobalConfig().TempStoragePath, "oom_sql"+time.Now().Format(time.RFC3339))
+	eqh.record.lastLogFileName = append(eqh.record.lastLogFileName, fileName)
+	f, err := os.Create(fileName)
+	if err != nil {
+		logutil.BgLogger().Error("Create oom record file fail.", zap.Error(err))
+		return
+	}
+	defer func() {
+		err := f.Close()
+		if err != nil {
+			logutil.BgLogger().Error("Close oom record file fail.", zap.Error(err))
+		}
+	}()
+	printTop10 := func(cmp func(i, j int) bool) {
+		sort.Slice(pinfo, cmp)
+		list := pinfo
+		if len(list) > 10 {
+			list = list[:10]
+		}
+		var buf strings.Builder
+		for i, info := range list {
+			buf.WriteString(fmt.Sprintf("SQL %v: \n", i))
+			fields := genLogFields(now.Sub(info.Time), info)
+			for _, field := range fields {
+				switch field.Type {
+				case zapcore.StringType:
+					buf.WriteString(fmt.Sprintf("%v: %v", field.Key, field.String))
+				case zapcore.Uint8Type, zapcore.Uint16Type, zapcore.Uint32Type, zapcore.Uint64Type:
+					buf.WriteString(fmt.Sprintf("%v: %v", field.Key, uint64(field.Integer)))
+				case zapcore.Int8Type, zapcore.Int16Type, zapcore.Int32Type, zapcore.Int64Type:
+					buf.WriteString(fmt.Sprintf("%v: %v", field.Key, field.Integer))
+				}
+				buf.WriteString("\n")
+			}
+		}
+		_, err = f.WriteString(buf.String())
+	}
+
+	_, err = f.WriteString("Top 10 memory usage of SQL for OOM analyze\n")
+	printTop10(func(i, j int) bool {
+		return pinfo[i].StmtCtx.MemTracker.MaxConsumed() > pinfo[j].StmtCtx.MemTracker.MaxConsumed()
+	})
+
+	_, err = f.WriteString("Top 10 time usage of SQL for OOM analyze\n")
+	printTop10(func(i, j int) bool {
+		return pinfo[i].Time.Before(pinfo[j].Time)
+	})
+
+	logutil.BgLogger().Info("Get oom sql successfully.", zap.Any("SQLs file path:", eqh.record.lastLogFileName))
```

Comment:
1. record SQLs with the most memory usage or time usage successfully
2. file path
3. eqh.record.lastLogFileName --> fileName?